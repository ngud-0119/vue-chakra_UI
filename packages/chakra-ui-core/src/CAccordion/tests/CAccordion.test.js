import { CAccordion, CAccordionItem, CAccordionHeader, CAccordionPanel, CAccordionIcon } from '..'
import { render, userEvent, fireEvent } from '@/tests/test-utils'

const renderComponent = (props) => {
  const base = {
    components: { CAccordion, CAccordionItem, CAccordionHeader, CAccordionPanel, CAccordionIcon },
    ...props
  }
  return render(base)
}

it('should render correctly', () => {
  const { asFragment } = renderComponent(
    {
      template: `
    <CAccordion>
      <CAccordionItem id="ac-1">
        <CAccordionHeader>Section 1 title<CAccordionIcon /></CAccordionHeader>
        <CAccordionPanel>Section 1 text</CAccordionPanel>
      </CAccordionItem>
    </CAccordion>
    `
    }
  )
  expect(asFragment()).toMatchSnapshot()
})

it('uncontrolled: It opens the accordion panel', () => {
  const { getByTestId } = renderComponent({
    template: `
    <CAccordion>
      <CAccordionItem>
        <CAccordionHeader data-testid="button">Section 1 title<CAccordionIcon /></CAccordionHeader>
        <CAccordionPanel data-testid="panel">Section 1 text</CAccordionPanel>
      </CAccordionItem>
    </CAccordion>
    `
  }
  )
  const button = getByTestId('button')
  expect(button).toHaveAttribute('aria-expanded', 'true')
})

it('uncontrolled: toggles the accordion on click', () => {
  const { getByText } = renderComponent({
    template: `
    <CAccordion>
      <CAccordionItem>
        <CAccordionHeader>Trigger</CAccordionHeader>
        <CAccordionPanel>Panel</CAccordionPanel>
      </CAccordionItem>
    </CAccordion>`
  })

  const trigger = getByText('Trigger')

  userEvent.click(trigger)
  expect(trigger).toHaveAttribute('aria-expanded', 'true')

  // you can't toggle an accordion without passing `allowToggle`
  userEvent.click(trigger)
  expect(trigger).toHaveAttribute('aria-expanded', 'true')
})

// test the only one accordion can be visible + is not togglable
it('only one accordion can be visible + is not togglable', () => {
  const { getByText } = renderComponent({
    template: `
    <CAccordion>
      <CAccordionItem>
        <CAccordionHeader>First section</CAccordionHeader>
        <CAccordionPanel>Panel 1</CAccordionPanel>
      </CAccordionItem>
      <CAccordionItem>
        <CAccordionHeader>Second section</CAccordionHeader>
        <CAccordionPanel>Panel 2</CAccordionPanel>
      </CAccordionItem>
    </CAccordion>`
  })

  const firstAccordion = getByText('First section')

  userEvent.click(firstAccordion)
  expect(firstAccordion).toHaveAttribute('aria-expanded', 'true')

  userEvent.click(firstAccordion)
  expect(firstAccordion).toHaveAttribute('aria-expanded', 'true')
})

// test the only one accordion can be visible + is togglable
it('only one accordion can be visible + is togglable', () => {
  const { getByText } = renderComponent({
    template: `
    <CAccordion allowToggle>
      <CAccordionItem>
        <CAccordionHeader>First section</CAccordionHeader>
        <CAccordionPanel>Panel 1</CAccordionPanel>
      </CAccordionItem>
      <CAccordionItem>
        <CAccordionHeader>Second section</CAccordionHeader>
        <CAccordionPanel>Panel 2</CAccordionPanel>
      </CAccordionItem>
    </CAccordion>`
  })
  const firstAccordion = getByText('First section')

  // TODO: Why its not working?
  // fireEvent.click(firstAccordion)
  // expect(firstAccordion).toHaveAttribute('aria-expanded', 'false')

  userEvent.click(firstAccordion)
  expect(firstAccordion).toHaveAttribute('aria-expanded', 'true')
})

// test that multiple accordions can be opened + is togglable
it('multiple accordions can be opened + is togglable', () => {
  const { getByText } = renderComponent({
    template: `
    <CAccordion allowMultiple :defaultIndex="[0]">
      <CAccordionItem>
        <CAccordionHeader>First section</CAccordionHeader>
        <CAccordionPanel>Panel 1</CAccordionPanel>
      </CAccordionItem>
      <CAccordionItem>
        <CAccordionHeader>Second section</CAccordionHeader>
        <CAccordionPanel>Panel 2</CAccordionPanel>
      </CAccordionItem>
    </CAccordion>`
  })

  const firstAccordion = getByText('First section')
  const secondAccordion = getByText('Second section')

  userEvent.click(firstAccordion)
  expect(firstAccordion).toHaveAttribute('aria-expanded', 'true')

  userEvent.click(secondAccordion)
  expect(firstAccordion).toHaveAttribute('aria-expanded', 'true')
})

// it has the proper aria attributes
it('has the proper aria attributes', () => {
  const { getByText, getByTestId } = renderComponent({
    template: `
    <CAccordion>
      <CAccordionItem>
        <CAccordionHeader>Section 1 title</CAccordionHeader>
        <CAccordionPanel data-testid="panel1">Panel 1</CAccordionPanel>
      </CAccordionItem>
    </CAccordion>`
  })
  const button = getByText('Section 1 title')
  const panel = getByTestId('panel1')
  expect(button).toHaveAttribute('aria-controls')
  expect(button).toHaveAttribute('aria-expanded')
  expect(panel).toHaveAttribute('aria-labelledby')
})

// test that tab moves focus to the next focusable element
it('tab moves focus to the next focusable element', () => {
  const { getByText } = renderComponent({
    template: `
    <CAccordion>
      <CAccordionItem>
        <CAccordionHeader>First section</CAccordionHeader>
        <CAccordionPanel>Panel 1</CAccordionPanel>
      </CAccordionItem>
      <CAccordionItem>
        <CAccordionHeader>Second section</CAccordionHeader>
        <CAccordionPanel>Panel 2</CAccordionPanel>
      </CAccordionItem>
      <CAccordionItem>
        <CAccordionHeader>Last section</CAccordionHeader>
        <CAccordionPanel>Panel 3</CAccordionPanel>
      </CAccordionItem>
    </CAccordion>`
  })

  const first = getByText('First section')
  const second = getByText('Second section')
  const last = getByText('Last section')

  userEvent.tab()
  expect(first).toHaveFocus()

  userEvent.tab()
  expect(second).toHaveFocus()

  userEvent.tab()
  expect(last).toHaveFocus()
})

// test that aria-contols for button is same as id for panel
it('aria-contols for button is same as id for panel', () => {
  const { getByText, getByTestId } = renderComponent({
    template: `
    <CAccordion>
      <CAccordionItem>
        <CAccordionHeader>Section 1 title</CAccordionHeader>
        <CAccordionPanel data-testid="panel1">Panel 1</CAccordionPanel>
      </CAccordionItem>
    </CAccordion>`
  })

  const button = getByText('Section 1 title')
  const panel = getByTestId('panel1')
  expect(button.getAttribute('aria-controls')).toEqual(panel.getAttribute('id'))
})

// test that aria-expanded is true/false when accordion is open/closed
it('aria-expanded is true/false when accordion is open/closed', () => {
  const { getByText } = renderComponent({
    template: `
    <CAccordion :defaultIndex="[0]">
      <CAccordionItem>
        <CAccordionHeader>Section 1 title</CAccordionHeader>
        <CAccordionPanel>Panel 1</CAccordionPanel>
      </CAccordionItem>
      <CAccordionItem>
        <CAccordionHeader>Section 2 title</CAccordionHeader>
        <CAccordionPanel>Panel 2</CAccordionPanel>
      </CAccordionItem>
    </CAccordion>`
  })

  const button = getByText('Section 1 title')
  expect(button).toHaveAttribute('aria-expanded', 'true')
})

// test that panel has role=region and aria-labelledby
it('panel has role=region and aria-labelledby', () => {
  const { getByTestId } = renderComponent({
    template: `
    <CAccordion>
      <CAccordionItem>
        <CAccordionHeader>Section 1 title</CAccordionHeader>
        <CAccordionPanel data-testid="panel1">Panel 1</CAccordionPanel>
      </CAccordionItem>
    </CAccordion>`
  })
  const panel = getByTestId('panel1')

  expect(panel).toHaveAttribute('aria-labelledby')
  expect(panel).toHaveAttribute('role', 'region')
})

// eslint-disable-next-line no-undef
xit('arrow up & down moves focus to next/previous accordion', () => {
  const { getByText } = renderComponent({
    template: `
    <CAccordion>
      <CAccordionItem>
        <CAccordionHeader>Section 1 title</CAccordionHeader>
        <CAccordionPanel>Panel 1</CAccordionPanel>
      </CAccordionItem>
      <CAccordionItem>
        <CAccordionHeader>Section 2 title</CAccordionHeader>
        <CAccordionPanel>Panel 2</CAccordionPanel>
      </CAccordionItem>
    </CAccordion>`
  })

  const first = getByText('Section 1 title')
  const second = getByText('Section 2 title')

  fireEvent.keyDown(first, { key: 'ArrowDown', keyCode: 40 })
  expect(second).toHaveFocus()

  fireEvent.keyDown(second, { key: 'ArrowUp', keyCode: 38 })
  expect(first).toHaveFocus()
})
