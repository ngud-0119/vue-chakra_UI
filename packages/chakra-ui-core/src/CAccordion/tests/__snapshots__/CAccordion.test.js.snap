// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<DocumentFragment>
  .emotion-2 {
  border-top-width: 1px;
}

.emotion-2:last-of-type {
  border-bottom-width: 1px;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  outline: 2px solid transparent;
  outline-offset: 2px;
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
  -webkit-padding-start: var(--chakra-space-4);
  padding-inline-start: var(--chakra-space-4);
  -webkit-padding-end: var(--chakra-space-4);
  padding-inline-end: var(--chakra-space-4);
  padding-top: var(--chakra-space-2);
  padding-bottom: var(--chakra-space-2);
}

.emotion-3:focus,
.emotion-3[data-focus] {
  box-shadow: var(--chakra-shadows-outline);
}

.emotion-3:hover,
.emotion-3[data-hover] {
  background: var(--chakra-colors-blackAlpha-50);
}

.emotion-3[disabled],
.emotion-3[aria-disabled=true],
.emotion-3[data-disabled] {
  opacity: 0.4;
  cursor: not-allowed;
}

.emotion-4 {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-backface-visibility: hidden;
  backface-visibility: hidden;
}

.emotion-4:not(:root) {
  overflow: hidden;
}

.emotion-5 {
  width: 1em;
  height: 1em;
  display: inline-block;
  vertical-align: middle;
}

.emotion-7 {
  opacity: 1;
  -webkit-transform: rotate(-180deg);
  -moz-transform: rotate(-180deg);
  -ms-transform: rotate(-180deg);
  transform: rotate(-180deg);
  -webkit-transition: -webkit-transform 0.2s;
  transition: transform 0.2s;
  transform-origin: center;
}

.emotion-8 {
  padding-top: var(--chakra-space-2);
  -webkit-padding-start: var(--chakra-space-4);
  padding-inline-start: var(--chakra-space-4);
  -webkit-padding-end: var(--chakra-space-4);
  padding-inline-end: var(--chakra-space-4);
  padding-bottom: var(--chakra-space-5);
}

<div
    class="emotion-0"
    data-chakra-component="CAccordion"
  >
    <div
      class="emotion-0 emotion-2"
      data-chakra-component="CAccordionItem"
    >
      <button
        aria-controls="accordion-panel-ac-1"
        aria-expanded="true"
        class="emotion-3"
        data-chakra-component="CAccordionHeader"
        id="accordion-header-ac-1"
        type="button"
      >
        Section 1 title
        <svg
          class="emotion-4 emotion-5 emotion-0 emotion-7"
          data-chakra-component="CAccordionIcon"
          role="presentation"
          viewBox="0 0 24 24"
        >
          
    
          <path
            d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
            fill="currentColor"
          />
        </svg>
      </button>
       
      <div
        data-chakra-component="CCollapse"
        style="overflow: hidden; visibility: hidden; height: 0px;"
      >
        <div
          aria-labelledby="accordion-header-ac-1"
          class="emotion-8"
          data-chakra-component="CAccordionPanel"
          id="accordion-panel-ac-1"
          role="region"
        >
          Section 1 text
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
